torch/distributed/rpc/backend_registry.py

```diff
@@ -353,6 +353,15 @@ def _tensorpipe_init_backend_handler(store, name, rank, world_size, rpc_backend_
 
     api._init_rpc_states(agent)
 
+    # Run one dummy round of RPC to initialize channels/transports. Without
+    # this, it's easy to hit timeout in rpc.shutdown() if there is no other RPC
+    # on that process before rpc.shutdown(), as the agent initialization can
+    # take longer than 5s.
+    api._all_gather(None, timeout=rpc_constants.DEFAULT_RPC_TIMEOUT_SEC)
+    # Need a barrier here to make sure no peers leave before the rank0 finishes
+    # _all_gather
+    group.barrier().wait()
```

Comment:
I'm wondering: would this also work if we only did the barrier, _without_ the `_all_gather` call? That way we would prevent any worker from advancing until all workers have finished initializing their agents.

Reply:
No, after removing that `all_gather`, I still hit the following error. And this is because, without running RPC here, some ranks' `shutdown()` will be the first place to launch RPC. And the first RPC on one rank is very slow due to initialization delay, and will lead to the timeout error.  

```
Failed to respond to 'Shutdown Proceed' in time, got error RPCErr:1:RPC ran for more than set timeout (5000 ms) and will now be marked with an error
Failed to respond to 'Shutdown Proceed' in time, got error RPCErr:1:RPC ran for more than set timeout (5000 ms) and will now be marked with an error
```